NEW General purpose search form component

maxime-rainville opened the pull request:
# Parent issue
* https://github.com/silverstripe/silverstripe-cms/issues/2184

lukereative commented on client/src/components/Search/SearchToggle.js:
```diff
@@ -0,0 +1,34 @@
+/**
+ * Simple Search toggle than can be use to enable/disable a Search component.
+ */
+const SearchToggle = ({ onToggle, toggled }) => (<button
```
Use reactstrap `Button`

maxime-rainville replied:
Done

lukereative commented on client/src/components/Search/SearchForm.js:
```diff
@@ -0,0 +1,54 @@
+ * Display an advanced filter form.
+ */
+const SearchForm = ({ expanded, onSearch, onClear, formSchemaUrl, id }) => (
+  <Collapse id={id} className="search-form" isOpen={expanded}>
```
This should have a nested `div` (ie. `search__filter__form`) with the padding applied to that rather than the Collapse element itself, this fixes the jumpy sliding animation

maxime-rainville replied:
Done ... so pretty to look at it slide now.

lukereative replied:
Yea, it was hurting my eyes before 😵

lukereative commented on client/src/components/Search/SearchForm.js:
```diff
@@ -0,0 +1,54 @@
+const SearchForm = ({ expanded, onSearch, onClear, formSchemaUrl, id }) => (
+  <Collapse id={id} className="search-form" isOpen={expanded}>
+    {formSchemaUrl && <FormBuilderLoader identifier={identifier} schemaUrl={formSchemaUrl} />}
+    <button className={searchButtonClasses} onClick={onSearch}>{searchLabel}</button>
```
Reactstrap `Button`

maxime-rainville replied:
Done

lukereative commented on client/src/components/Search/SearchForm.js:
```diff
@@ -0,0 +1,54 @@
+  <Collapse id={id} className="search-form" isOpen={expanded}>
+    {formSchemaUrl && <FormBuilderLoader identifier={identifier} schemaUrl={formSchemaUrl} />}
+    <button className={searchButtonClasses} onClick={onSearch}>{searchLabel}</button>
+    <button className={clearButtonClasses} onClick={onClear}>{clearLabel}</button>
```
Reactstrap `Button`

maxime-rainville replied:
Done

lukereative commented on client/src/components/Search/SearchBox.scss:
```diff
@@ -0,0 +1,192 @@
+  border-top: 0;
+  border-bottom: 0;
+  border-radius: 0;
+  font-size: 15px;
```
This should probably use $font-size-lg (16px) rather than a custom one

maxime-rainville replied:
Done.

lukereative commented on client/src/components/Search/SearchBox.scss:
```diff
@@ -0,0 +1,192 @@
+  // Rotate ^ when filter-panel is open
+  &::before {
+    margin: 0;
+    transform: rotate(180deg) translateY(2px);
```
You need to apply `display: inline-block;` for this transformation to work, also might be nice to use `.5turn` (`turn` is a little known valid unit) to make it a little bit clearer as to what the transformation is.

maxime-rainville replied:
That rotation style is a left over from the Asset Search component. The mock up doesn't actually have it. It might just be an omission. I'll need to have a chat with @clarkepaul to see what he wants.

lukereative commented on client/src/components/Search/SearchBox.scss:
```diff
@@ -0,0 +1,192 @@
+  }
+
+  &.collapsed::before {
+    transform: rotate(0deg) translateY(0);
```
Remove the `deg` unit from `0`

lukereative commented on client/src/components/Search/SearchBox.scss:
```diff
@@ -0,0 +1,192 @@
+    width: 100%;
+  }
+
+  .font-icon-search {
```
You probably should chuck `pointer-events: none;` on this so it doesn't prevent clicking into the field.

maxime-rainville replied:
Done. I've applied to the _enter_ bit as well.

lukereative commented on client/src/components/Search/SearchBox.scss:
```diff
@@ -0,0 +1,192 @@
+  top: 16px;
+  padding: 0 4px;
+  font-size: $font-size-sm;
+  border-radius: 2px;
```
Use `$border-radius-sm`

maxime-rainville replied:
Done.

maxime-rainville commented on client/src/components/Search/Search.js:
```diff
@@ -0,0 +1,361 @@
+   */
+  open() {
+    this.show();
+    setTimeout(this.focusInput, 50);
```
That got copied over from the original search component in asset-admin. I think the timeout is there because there's an animation when the filters form get toggled. So if you try to set the focus right away, the field won't exist yet.

I'll see if there's an easy way to use a callback instead.

maxime-rainville replied:
I had a look and the timeout is indeed there because immediately after the how is called, the form field won't be available yet. So there's nothing to focus on. The previous method is just state update, so it doesn't have a nice elegant callback I can hook into.

There might be a way to use a proper callback to handle the focus, but it's probably going to make the code more complex. So I'll leave it like that.

maxime-rainville commented on client/src/components/Search/Search.js:
```diff
@@ -0,0 +1,361 @@
+    switch (this.state.display) {
+      case DISPLAY.VISIBLE:
+        this.expand();
+        setTimeout(this.focusFirstFormField, 50);
```
Ditto to the other one ... I'll have a look if there's a more elegant of doing this.

maxime-rainville replied:
Same as https://github.com/silverstripe/silverstripe-admin/pull/572#discussion_r204563191

maxime-rainville commented on client/src/components/Search/SearchBox.js:
```diff
@@ -0,0 +1,133 @@
+            autoFocus
+          />
+          <div className="icon font-icon-search" />
+          { (showFilters) && <button
```
Done.

maxime-rainville commented on client/src/components/Search/SearchBox.scss:
```diff
@@ -0,0 +1,192 @@
+    position: relative;
+    right: auto;
+
+    &.search-box__active {
```
I'll get rid of the active class altogher. The previous component was just hidding the search with CSS. I'm not rendering it at all when its not needed. So it's kind of redundant to have an active class here.

maxime-rainville commented on client/src/components/Search/SearchToggle.scss:
```diff
@@ -0,0 +1,9 @@
+.search-toggle {
+  z-index: 1;
+
+  &.search-toggle__active {
```
Done.

maxime-rainville commented on client/src/components/Search/Search.js:
```diff
@@ -0,0 +1,361 @@
+
+    // Decide if we display the X button
+    const hideable =
+      [BEHAVIOR.HIDEABLE, BEHAVIOR.TOGGLABLE].indexOf(displayBehavior) > -1;
```
Done

maxime-rainville commented on client/src/components/Search/Search.js:
```diff
@@ -0,0 +1,361 @@
+    if (this.state.display === DISPLAY.NONE) {
+      return (displayBehavior === BEHAVIOR.TOGGLABLE) ?
+        <SearchToggle onToggle={this.show} /> :
+        <div />;
```
Done

maxime-rainville commented on client/src/components/Search/SearchBox.js:
```diff
@@ -0,0 +1,133 @@
+
+    // Build classes
+    const searchClasses = ['search-box', 'search-box__active'];
+    const advancedButtonClasses = [
```
Done.

maxime-rainville commented on client/src/components/Search/SearchBox.js:
```diff
@@ -0,0 +1,133 @@
+            className={advancedButtonClasses.join(' ')}
+            title={i18n._t('Admin.ADVANCED', 'Advanced')}
+          />}
+          { hideable && <button
```
Switched to Button.

maxime-rainville commented:
I've got most of it sorted.

## Actioned points
* Placeholder text should be lighter.
* Dropdown button should have a darkened hover and active state (should be fixed by using reactstrap Button)
*  'X' icon needs to be bigger and the same icon colour as the others
* Should have a pattern library example of a wide search bar (_I had to froce the width at some arbitrary number_)
* The advanced filters width should be ~320px rather than 448px to match the design
* For vertically centring absolutely positioned elements try use

## Other points

### Search icon should be 16px not 20px and a lighter colour
I've reduce the size to match the mockup. The currently colour in the mock up matches exactly what's being used however.

### Dropdown button caret should be aligned same distance from left side as the search button to the right
As far as I can tell, it is.

### The 'enter' element is too far in when there is no advanced filters
Couldn't replicate this one.

maxime-rainville commented:
I think the build is failing right now because it's not using the most up-to-date version of our webpack config package.

There's a separate PR for that https://github.com/silverstripe/webpack-config/pull/23

If we can publish an updated webpack-config, I'll be able to update our `yarn.lock` file which should make the build for this PR green.

lukereative commented on client/src/components/Search/SearchBox.js:
```diff
@@ -44,30 +45,31 @@ class SearchBox extends Component {
             aria-labelledby={`${id}_label`}
             type="text"
-            name="name"
+            name={this.props.name}
```
This should probably be a part of the destructuring of `this.props` above

maxime-rainville replied:
Done.

maxime-rainville commented:
* The search icon has been set to block.
* I've adjusted the width of the `Enter` hint to allow it to be bigger. I've added a tiny bit more padding to the Input to accommodate variation of length translation for `Enter`.
* I never was quite sure what to do with the `Clear` button. It would clash with the [Enter]. The only thing I could think is that the Clear button should be hidden when the input has the focus and reappear when the input loses focus. I'll add it to my list of points to discuss with Paul.

maxime-rainville commented:
@lukereative You can have another go at this guy. The main change is the addition of logic to track whatever to search component is _dirty_ or not. This controls whatever the enter hint or the clear button gets displayed in the searchbox.

Triggering doSearch, should reset dirty to false.

maxime-rainville commented on client/src/components/Search/SearchBox.js:
```diff
@@ -0,0 +1,154 @@
+          >{i18n._t('Admin.CLEAR', 'Clear')}</Button>}
+          {children}
+          <div className="icon font-icon-search" />
+          { (showFilters) && <Button
```
Done

maxime-rainville commented on client/src/components/Search/SearchBox.js:
```diff
@@ -0,0 +1,154 @@
+        </div>
+      </div>
+    );
+  }
```
Done

maxime-rainville commented on client/src/components/Search/SearchToggle.js:
```diff
@@ -0,0 +1,33 @@
+const SearchToggle = ({ onToggle, toggled }) => (<Button
+  title={label}
+  onClick={onToggle}
+  className={btnClasses + (toggled ? ' search-toggle__active' : '')}
```
Done. I also added a specific entry in the pattern lib for the `SearchToggle` since it can be used independently of the `Search` component.
